File: mobile/lib/models/transaction.dart
enum TransactionStatus {
  created,
  initiated,
  pending,
  success,
  failed,
  cancelled,
  timeout,
  unknown;

  static TransactionStatus fromString(String s) {
    return TransactionStatus.values.firstWhere(
      (e) => e.name.toUpperCase() == s.toUpperCase(),
      orElse: () => TransactionStatus.unknown,
    );
  }

  bool get isTerminal =>
      this == success || this == failed || this == cancelled || this == timeout;

  bool get isPending =>
      this == created || this == initiated || this == pending;
}

enum TransactionType {
  stkPush,
  c2b,
  b2c,
  b2b;

  static TransactionType fromString(String s) {
    const map = {
      'STK_PUSH': TransactionType.stkPush,
      'C2B': TransactionType.c2b,
      'B2C': TransactionType.b2c,
      'B2B': TransactionType.b2b,
    };
    return map[s.toUpperCase()] ?? TransactionType.stkPush;
  }

  String get displayName {
    switch (this) {
      case stkPush: return 'Payment';
      case c2b: return 'Received';
      case b2c: return 'Sent';
      case b2b: return 'Business';
    }
  }
}

class Transaction {
  final String id;
  final String clientReference;
  final String? providerReference;
  final String phoneNumber;
  final double amount;
  final String currency;
  final TransactionType type;
  final TransactionStatus status;
  final String accountReference;
  final String description;
  final String? providerResponseDescription;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? completedAt;

  const Transaction({
    required this.id,
    required this.clientReference,
    this.providerReference,
    required this.phoneNumber,
    required this.amount,
    required this.currency,
    required this.type,
    required this.status,
    required this.accountReference,
    required this.description,
    this.providerResponseDescription,
    required this.createdAt,
    required this.updatedAt,
    this.completedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String,
      clientReference: json['client_reference'] as String,
      providerReference: json['provider_reference'] as String?,
      phoneNumber: json['phone_number'] as String,
      amount: (json['amount'] as num).toDouble(),
      currency: json['currency'] as String? ?? 'KES',
      type: TransactionType.fromString(json['transaction_type'] as String),
      status: TransactionStatus.fromString(json['status'] as String),
      accountReference: json['account_reference'] as String,
      description: json['description'] as String,
      providerResponseDescription: json['provider_response_description'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
    );
  }
}

File: backend/app/models/transaction.py
import enum
import uuid
from datetime import datetime, timezone
from sqlalchemy import DateTime, Enum, Index, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column
from app.db.database import Base


class TransactionStatus(str, enum.Enum):
    CREATED = "CREATED"
    INITIATED = "INITIATED"
    PENDING = "PENDING"
    SUCCESS = "SUCCESS"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"
    TIMEOUT = "TIMEOUT"
    UNKNOWN = "UNKNOWN"


class TransactionType(str, enum.Enum):
    STK_PUSH = "STK_PUSH"
    C2B = "C2B"
    B2C = "B2C"
    B2B = "B2B"


def _now() -> datetime:
    return datetime.now(timezone.utc)


class Transaction(Base):
    __tablename__ = "transactions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    idempotency_key: Mapped[str | None] = mapped_column(String(128), unique=True, nullable=True)

    # Client-facing reference
    client_reference: Mapped[str] = mapped_column(String(64), index=True)
    # Safaricom-assigned reference (MpesaReceiptNumber etc.)
    provider_reference: Mapped[str | None] = mapped_column(String(64), nullable=True, index=True)
    # CheckoutRequestID from STK push
    checkout_request_id: Mapped[str | None] = mapped_column(String(128), nullable=True, unique=True)

    phone_number: Mapped[str] = mapped_column(String(20))
    amount: Mapped[float] = mapped_column(Numeric(12, 2))
    currency: Mapped[str] = mapped_column(String(3), default="KES")

    transaction_type: Mapped[TransactionType] = mapped_column(Enum(TransactionType))
    status: Mapped[TransactionStatus] = mapped_column(
        Enum(TransactionStatus), default=TransactionStatus.CREATED, index=True
    )

    account_reference: Mapped[str] = mapped_column(String(12))
    description: Mapped[str] = mapped_column(String(13))

    provider_response_code: Mapped[str | None] = mapped_column(String(10), nullable=True)
    provider_response_description: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now, index=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now, onupdate=_now)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    __table_args__ = (
        Index("ix_transactions_created_at", "created_at"),
    )
